Validate proxy taken from env variables and allow only absolute paths for config proxy-ca-file

anjannath opened the pull request:
- validate proxy urls taken from env vairbles
- validate that proxy certificate file exists before trying to read it
- only allow absolute paths for the `proxy-ca-file` config setting

Fixes #3690 #3098 #3641

cfergeau commented:
Can `config: make proxy-ca-file only accept absolute paths` cause existing configs to stop working?
Would it be possible to convert relative to absolute and to use absolute throughout crc's codebase? This would be more user-friendly.

anjannath commented:
> Can `config: make proxy-ca-file only accept absolute paths` cause existing configs to stop working? 

no, existing configs should keep working, the stricter validation only happens when user configures a value via `crc config set`, when retrieving the value for use by other parts in the codebase the validation will not run

> Would it be possible to convert relative to absolute and to use absolute throughout crc's codebase? This would be more user-friendly.

yeah it'll better, but this would mean we need to mutate the config value after user has provided it, currently in the config pkg we don't have such provisions, i was thinking to follow the same pattern as the _change notifiers_ (and call it _mutators_) did you have another idea on how to implement this?

cfergeau committed:
config: Add config Path type and use it for proxy-ca-file

This allows to convert relative paths to absolute paths at the time
the config value is set.

fixes #3641

cfergeau commented:
> > Would it be possible to convert relative to absolute and to use absolute throughout crc's codebase? This would be more user-friendly.
> 
> yeah it'll better, but this would mean we need to mutate the config value after user has provided it, currently in the config pkg we don't have such provisions, i was thinking to follow the same pattern as the _change notifiers_ (and call it _mutators_) did you have another idea on how to implement this?

I had https://github.com/cfergeau/crc/commit/2d5e48887ba1a537bd8ab949e9283f31e8898beb in the back of my mind, which seems to work as expected after some quick prototyping (only test I did is what is in the unit test)

anjannath commented:
> I had [cfergeau@2d5e488](https://github.com/cfergeau/crc/commit/2d5e48887ba1a537bd8ab949e9283f31e8898beb) in the back of my mind, which seems to work as expected after some quick prototyping (only test I did is what is in the unit test)

i was thinking to implement it like this: https://github.com/anjannath/crc/commit/5d87b1dff4328194939d2901490f46f702b536b8 but with the `Path` type i think its better, as while adding the config setting its clear, will take your commit do some more testing and update the PR, thanks!

cfergeau commented on pkg/crc/network/httpproxy/proxy.go:
```diff
@@ -66,10 +68,14 @@ func NewProxyDefaults(httpProxy, httpsProxy, noProxy, proxyCAFile string) (*Prox
 	if DefaultProxy.HTTPProxy == "" {
-		DefaultProxy.HTTPProxy = envProxy.HTTPProxy
+		if err := ValidateProxyURL(envProxy.HTTPProxy, false); err == nil {
+			DefaultProxy.HTTPProxy = envProxy.HTTPProxy
```
When there is an error, I'd add a warning so that it's clear proxy from the environment is being ignored.

anjannath replied:
added a warning msg

cfergeau commented on pkg/crc/network/httpproxy/proxy.go:
```diff
@@ -45,6 +46,9 @@ func readProxyCAData(proxyCAFile string) (string, error) {
 		return "", nil
 	}
+	if !crcos.FileExists(proxyCAFile) {
+		return "", nil
```
Is there a validation happening later on? Ideally, most crc commands would not fail when the proxy config is invalid. It's probably only `crc config`, `crc setup` and `crc start` which should report a hard failure when the proxy is invalid, the other ones are not dependent on the proxy settings.

anjannath replied:
> Is there a validation happening later on? 

no, there's validation when running `config set`. and then here when proxy is being setup which is very early on in `root.go`.  (we send telemetry data from each command, and that uses the proxy settings)

> It's probably only crc config, crc setup and crc start which should report a hard failure when the proxy is invalid, the other ones are not dependent on the proxy settings.

maybe for these commands we can re-validate (add a "re-validation" helper in proxy package) the proxy settings and this will be hard failure if validation fails

anjannath replied:
Thinking about this again, doing this selective hard failure for `start` and `setup` seems not to be necessary, as we have added validation of proxy config from environment, now it should not be possible to configure invalid proxy settings at all via the environment or via `crc config set`

for the `proxy-ca-file` we have validation while `config set` is used to make sure the file exists, and with the conversion of relative to absolute paths, user shouldn't end up in a situation where they have invalid path to `proxy-ca-file`

now the only way invalid proxy settings can end up in the config is by manually editing the config file, which i think should be a hard failure for all the commands, WDYT?

cfergeau replied:
> for the proxy-ca-file we have validation while config set is used to make sure the file exists, and with the conversion of relative to absolute paths, user shouldn't end up in a situation where they have invalid path to proxy-ca-file

it's a user problem, but
```
$ crc config set proxy-ca-file foo.crt
$ rm foo.crt
```

> Thinking about this again, doing this selective hard failure for start and setup seems not to be necessary, as we have added validation of proxy config from environment, now it should not be possible to configure invalid proxy settings at all via the environment or via crc config set

We validate the proxy config, and if for example `HTTP_PROXY` is invalid in the environment, we ignore it right?
If the user made a typo when setting it, it's better to make this very visible than ignoring it, no?

cfergeau replied:
If there are already warnings for all of this every time crc starts up, it might be enough ,I'll have to check how it looks :)

anjannath replied:
> We validate the proxy config, and if for example HTTP_PROXY is invalid in the environment, we ignore it right?
If the user made a typo when setting it, it's better to make this very visible than ignoring it, no?

yes i changed that its not silently ignoring it now. If the configured `proxy-ca-file` path is not present, it returns an error which is a hard failure for the all the commands, also with a msg asking user to either unset or correct the config

cfergeau commented on pkg/crc/network/httpproxy/proxy.go:
```diff
@@ -41,7 +42,7 @@ func (p *ProxyConfig) String() string {
 
 func readProxyCAData(proxyCAFile string) (string, error) {
-	if proxyCAFile == "" {
+	if proxyCAFile == "" || !filepath.IsAbs(proxyCAFile) {
```
I'm not sure why we want to ignore the error in this situation?

cfergeau commented on pkg/crc/network/httpproxy/proxy.go:
```diff
@@ -75,11 +75,15 @@ func NewProxyDefaults(httpProxy, httpsProxy, noProxy, proxyCAFile string) (*Prox
 		if err := ValidateProxyURL(envProxy.HTTPSProxy, true); err == nil {
 			DefaultProxy.HTTPSProxy = envProxy.HTTPSProxy
+		} else {
+			logging.Warnf("HTTPS_PROXY environment variable is being ignored as it failed validation: %v", err)
```
Can these 2 warnings be moved to the first commit?

cfergeau commented on pkg/crc/network/httpproxy/proxy.go:
```diff
@@ -260,3 +264,16 @@ func HTTPTransport() http.RoundTripper {
+		return err
+	}
+	if proxyConfig.ProxyCAFile != "" {
+		if proxyConfig.ProxyCACert == "" {
```
The 2 commits before this one ignore various errors happening in `readProxyCAData`, this check for `ProxyCACert` is an indirect way of checking if we ignored any errors.
We also ignore invalid proxy settings from the environment, but this will not be detected here (not necessarily mandatory, but would be nice if everything could be reported to the user).

The commit logs are short on details about the bugs that are being fixed, so I'd need to debug this again to see if there are easy alternatives I can suggest to improve this :(

cfergeau reviewed (commented):
This PR ignores the errors which block startup, and then the code tries to figure out if an error was ignored to report it at the right time. Ideally we could do something more direct, but I'll need to look closer at it.

cfergeau commented:
These are these lines which are causing an error to happen very early in crc startup:
https://github.com/crc-org/crc/blob/85cc7183e4daf12b148b594a444e1d50546910e3/cmd/crc/cmd/root.go#L58-L60

Would be useful to mention it in one of the commit logs :)


In my opinion, if we want to ignore errors/output a warning, root.go is the place where we should do it, it's the only place calling NewProxyDefaults().

I think you can also add some kind of 'error' field to ProxyConfig, to record that there was an error reading the CA file, and/or that there was an error parsing the proxy from the env, ... Then it becomes easy to check at a later time if there are proxy config errors.
`start` has a call to `getProxyConfig()` relatively early during startup so an explicit "revalidateConfig" may not be needed there.

anjannath committed:
warn instead of fatal error when proxy configs are erroneous

when an error is returned from NewProxyConfig because of wrong
proxy env variables or missing proxy ca cert file, in root.go it
was fatal error which prevented running of any crc commands  and
user was not able to unset the invalid config

this changes it to be a warning, which will allow users to correct
the wrong config

this fixes #3690 #3098

anjannath committed:
mention about wrong proxy config via env variable in error msg

when proxy config is picked up from the env variables this modifies the
error message to inform user to check and correct the proxy urls in the
env variables

anjannath commented:
@cfergeau I've updated the PR with more descriptive commit msg and changed the ignoring of error to happen in root.go, also added a commit to let users know when invalid config is preset in the env variables, PTALA :)

cfergeau commented on pkg/crc/config/validations.go:
```diff
@@ -103,6 +104,17 @@ func validatePath(value interface{}) (bool, string) {
 	return true, ""
 }
 
+func validatePathAbs(value interface{}) (bool, string) {
```
This could be named `validateAbsolutePath`?

cfergeau reviewed (commented):
What happens if the user's config has a relative path to a CA file, and they upgrade to a version of crc with validateAbsolutePath? A warning will be printed? Will they still be able to use their relative path CA if it worked for them before?

anjannath commented:
> What happens if the user's config has a relative path to a CA file, and they upgrade to a version of crc with validateAbsolutePath? A warning will be printed? Will they still be able to use their relative path CA if it worked for them before?

yes, it'll keep working, since the validation only happens when user runs `crc config set proxy-ca-file` but there will be no warning also to ask them to migrate to absolute path..

anjannath commented:
I forgot about fc4dbfb4e504aaf439557f2cbadb16efd4e902c2 which is more user friendly than the `validaeAbsolutePath`

anjannath commented:
@cfergeau updated :)

cfergeau commented on pkg/crc/config/config.go:
```diff
@@ -239,11 +239,20 @@ func (c *Config) Get(key string) SettingValue {
 	}
 	return SettingValue{
 		Value:     value,
-		IsDefault: reflect.DeepEqual(setting.defaultValue, value),
```
Good catch!
There's a 
```
case Path, string:
		value = cast.ToString(value)
```
in this function, did you try changing it to:
```
case Path: value = Path(cast.ToString(value))
```
?

anjannath replied:
Yes tried this now, and it also works, updated to use this instead..

cfergeau commented on pkg/crc/config/config.go:
```diff
@@ -200,6 +211,7 @@ func (c *Config) Get(key string) SettingValue {
 		}
 	case string:
 		value = cast.ToString(value)
+	case Path:
```
Imo this should be `value = Path(cast.ToString(value))`

anjannath replied:
Doing this fails the existing test case, had to wrap the expected value in `Path` and `config view` was not showing the `proxy-ca-file` value for that to work had to also add a stringer method for the `Path` type

cfergeau commented on pkg/crc/config/settings_test.go:
```diff
@@ -158,3 +159,19 @@ func TestUnsetPreset(t *testing.T) {
+
+func TestPath(t *testing.T) {
+	cfg, err := newInMemoryConfig()
+	require.NoError(t, err)
```
A test for the default value would be useful imo as this was initially buggy.
Something like this should do the trick:

```
	assert.Equal(t, SettingValue{
		Value:     Path(""),
		Invalid:   false,
		IsDefault: true,
		IsSecret:  false,
	}, cfg.Get(ProxyCAFile))
```
